src/main/scala/com/github/sbt/duplicates/ClasspathEntity.scala

```diff
@@ -0,0 +1,81 @@
+package com.github.sbt.duplicates
+
+import java.nio.file.Files
+import java.security.{DigestInputStream, MessageDigest}
+
+import sbt._
+
+import scala.tools.nsc.interpreter.InputStream
+
+/**
+  * Lazy & comparable handle to a class or resource within a zip file or directory.
+  * Caches the CRC32.
+  * Falls back to reading the full file contents if absolutely necessary.
+  */
+sealed trait ClasspathEntity {
+
+  /**
+    * Name of the zip file or directory.
+    */
+  def source: File
+
+  /**
+    * Path within the zip file or directory.
+    *
+    * @example `scala/None$.class`
+    * @example `META-INF/MANIFEST.MF`
+    */
+  def name: String
+
+  /**
+    * The content of the class file or resource.
+    */
+  protected def withInputStream[T](f: InputStream => T): T
+
+  lazy val sha256: Seq[Byte] =
+    withInputStream { in =>
+      val dis = new DigestInputStream(in, MessageDigest.getInstance("SHA-256"))
+      IO.transfer(dis, NullOutputStream)
+      dis.getMessageDigest.digest().toSeq
+    }
```

Comment:
Relying on sbt's `IO` helpers, it could be simplified with:

```suggestion
  lazy val sha256: Seq[Byte] = {
    val bytes = withInputStream(Compat.IO.readBytes)
    MessageDigest.getInstance("SHA-256").digest(bytes).toSeq
  }
```

Reply:
Yours is simpler for sure.

One difference is that `readBytes` pulls the whole content into memory. The 20MB datadog jar is the biggest on my classpath and this section is run for my 27 projects in parallel. It's a bit of a stretch in my case, but that does add up to half of the default sbt heap over the worst case. Given there's no max projects or bound for jar size that this plugin may encounter, it seemed worth the extra lines to never have to worry about it again. WDYT? If you're still not convinced, I'll change.

If only we had an `inputStream.foldLeft(digest)((digest, chunk: Array[Byte]) => )`, the memory-safe solution would feel more natural.

Reply:
Indeed, completely missed that yours was lazily loading, which is indeed better, we can never know how fat an analyzed JAR would be.
Small suggestion though, to avoid using IO and introducing an OutputStream (reads all bytes and until the input is exhausted):

```suggestion
  lazy val sha256: Seq[Byte] =
    withInputStream { in =>
      val dis = new DigestInputStream(in, MessageDigest.getInstance("SHA-256"))
      while(dis.read() != -1) {}
      dis.getMessageDigest.digest().toSeq
    }
```

Reply:
Can do. I'll keep the `read(arr)` approach from `IO.transfer()`, since it's ~7x faster.

```
class DigestBench {

  private val bytes = Array.iterate[Byte](0.toByte, 8000)(b => (b + 1).toByte)
  private def stream() = new DigestInputStream(new BufferedInputStream(new ByteArrayInputStream(bytes)), MessageDigest.getInstance("SHA-256"))
  
  @Benchmark
  def readArr = {
    val in = stream()
    val buf = Array.ofDim[Byte](4000)
    while (in.read(buf) != -1) {}
    in.getMessageDigest.digest()
  }

  @Benchmark
  def readByte = {
    val in = stream()
    while (in.read() != -1) {}
    in.getMessageDigest.digest()
  }
}
```
```
[info] Benchmark              Mode  Cnt      Score       Error  Units
[info] DigestBench.readArr   thrpt    3  24096.342 ± 16817.796  ops/s
[info] DigestBench.readByte  thrpt    3   3458.116 ±   521.676  ops/s
```

Reply:
Good to know, that's quite a speedup.